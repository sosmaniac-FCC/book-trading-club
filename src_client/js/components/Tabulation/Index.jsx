import React from 'react';

import { Link } from 'react-router-dom';

const Index = (props) => (
    <li className="tab white-text">
        <Link onClick={props.handleClick} 
        className={
            ((index) => {
                if (index == 5) {
                    return "hide-on-small-only";
                }
                else {
                    return "";
                }
            })(props.i)}
        id={"option-" + props.i}
        to={((index) => {
                switch (index) {
                    case 0: return "/Book/BookSpread";
                    case 1: return "/User/" + props.username + "/Books";
                    case 2: return "/User/" + props.username + "/Trades";
                    case 3: return "/User/" + props.username + "/Settings";
                    case 4: return "/Login";
                    case 5: return "/";
                    default: throw "ERROR! -> render() of Tabulation computes indexOutOfBounds A";
                }
            })(props.i)}>
            {/* innerHTML */}
            {((index) => {
                switch (index) {
                    case 0: return "ALL BOOKS";
                    case 1: return "MY BOOKS";
                    case 2: return "TRADES";
                    case 3: return "SETTINGS";
                    case 4: return "LOGIN HERE";
                    case 5: return "LOGOUT";
                    default: throw "ERROR! -> render() of Tabulation computes indexOutOfBounds B";
                }
            })(props.i)}
        </Link>
    </li>
);

export default Index;