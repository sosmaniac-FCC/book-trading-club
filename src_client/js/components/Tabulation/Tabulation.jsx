import React from 'react';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { v4 } from 'uuid';

import { authenticate } from '../../reducers/userReducer/userActions.jsx';
import { handleLogout } from '../../reducers/tabReducer/tabActions.jsx';
import Index from './Index.jsx';

/* global $ */

class Tabulation extends React.Component {
    constructor(props) {
        super(props);
        
        this.handleClick = this.handleClick.bind(this);
    }
    
    handleClick({ target }) {
        // utilize regex to convert href back to its relative pathname
        const relativeURL = target.href.replace(/.*(\.io|\.com|\.gov|\.net|\.edu)/i, '');
        
        this.props.history.push(relativeURL);
        
        if (target.id == "option-5") {
            this.props.handleLogout();
            this.props.authenticate(null);
        }
    }
    
    // tabs are initialized automatically, but require reinitialization upon dynamic update
    componentDidUpdate() {
        $(document).ready(() => {
            $("ul.tabs").tabs();
        });
    }
    
    render() {
        // renders annoying-to-modify, deterministic tabulation
        const content = Object.keys(this.props).map(key => this.props[key]).reduce((accum, curr) => {
            // booleans indicate potential navigation bar options
            if (typeof curr == "boolean")
                accum.push(curr);
            return accum;
        }, []).map((item, i) => (
                item ? <Index i={i} handleClick={this.handleClick} username={this.props.username} key={v4()}/> : null
            )
        );
        
        return (
            <ul className="tabs tabs-fixed-width tabs-transparent">
                {content}
            </ul>
        );
    }
}

function mapStateToProps(state) {
    const { tabs } = state;
    const { users } = state;
    
    return {
        spread: tabs.spread,
        books: tabs.books,
        trades: tabs.trades,
        settings: tabs.settings,
        login: tabs.login,
        logout: tabs.logout,
        username: users.username,
        token: users.token
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        handleLogout,
        authenticate
    }, dispatch);
}

// withRouter used as a work-around due to materialize tabs muting native functionality
// of the <Link /> component
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Tabulation));

// componentWillMount() {
//     // ATTEMPT TO FIX THE INDICATOR REPOSITIONING ON PAGE RESIZE
//
//     window.addEventListener('resize', () => {
//         console.log("trigger resize");
//     });
// }