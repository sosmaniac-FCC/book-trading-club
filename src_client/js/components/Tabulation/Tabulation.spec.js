// expect from jest_expect.js
/* global expect */

import React from 'react';
import { shallow, configure } from 'enzyme';
import enzymeAdapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';

import { createStore } from 'redux';

import Tabulation from './Tabulation.jsx';

configure({ adapter: new enzymeAdapter() });

test('Dynamic tab content loads as expected.', () => {
    const store = createStore(() => ({
        books: {},
        tabs: {},
        trades: {},
        users: {}
    }));
    
    const component = shallow(<Tabulation store={store} />);
    const tree = toJSON(component);
    expect(tree).toMatchSnapshot();
});

test('Dynamic tab content should have no style attributes.', () => {
    const store = createStore(() => ({
        books: {},
        tabs: {},
        trades: {},
        users: {}
    }));
    
    const component = shallow(<Tabulation store={store} />);
    expect(component.find('[style]')).toHaveLength(0);
});