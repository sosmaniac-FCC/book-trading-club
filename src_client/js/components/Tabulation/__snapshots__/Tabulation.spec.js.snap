// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic tab content loads as expected. 1`] = `
<Tabulation
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
